p2p/host/base/base_host.go

```diff
@@ -0,0 +1,255 @@
+package base
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/libp2p/go-libp2p/p2p/host/pstoremanager"
+
+	"github.com/libp2p/go-libp2p-core/connmgr"
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-core/record"
+
+	"github.com/libp2p/go-eventbus"
+
+	logging "github.com/ipfs/go-log/v2"
+	ma "github.com/multiformats/go-multiaddr"
+	msmux "github.com/multiformats/go-multistream"
+)
+
+var log = logging.Logger("basehost")
+
+type Option func(*BaseHost) error
+
+func WithConnMgr(mgr connmgr.ConnManager) Option {
+	return func(h *BaseHost) error {
+		h.connmgr = mgr
+		return nil
+	}
+}
+
+func WithMultistreamMuxer(mux *msmux.MultistreamMuxer) Option {
+	return func(h *BaseHost) error {
+		h.mux = mux
+		return nil
+	}
+}
+
+type BaseHost struct {
+	network   network.Network
+	psManager *pstoremanager.PeerstoreManager
+	eventbus  event.Bus
+	mux       *msmux.MultistreamMuxer
+	connmgr   connmgr.ConnManager
+
+	closeOnce sync.Once
+
+	emitters struct {
+		evtLocalProtocolsUpdated    event.Emitter
+		evtPeerConnectednessChanged event.Emitter
+	}
+}
+
+var _ host.Host = &BaseHost{}
+
+func NewHost(n network.Network, opts ...Option) (*BaseHost, error) {
+	eventBus := eventbus.NewBus()
+	psManager, err := pstoremanager.NewPeerstoreManager(n.Peerstore(), eventBus)
+	if err != nil {
+		return nil, err
+	}
+	h := &BaseHost{
+		network:   n,
+		eventbus:  eventBus,
+		psManager: psManager,
+	}
+	for _, opt := range opts {
+		if err := opt(h); err != nil {
+			return nil, err
+		}
+	}
+	if h.mux == nil {
+		h.mux = msmux.NewMultistreamMuxer()
+	}
+	if h.connmgr == nil {
+		h.connmgr = &connmgr.NullConnMgr{}
+	} else {
+		n.Notify(h.connmgr.Notifee())
+	}
+
+	if h.emitters.evtLocalProtocolsUpdated, err = h.eventbus.Emitter(&event.EvtLocalProtocolsUpdated{}); err != nil {
+		return nil, err
+	}
+	h.emitters.evtPeerConnectednessChanged, err = h.eventbus.Emitter(&event.EvtPeerConnectednessChanged{})
+	if err != nil {
+		return nil, err
+	}
+	n.Notify(newPeerConnectWatcher(h.emitters.evtPeerConnectednessChanged))
+
+	if err := h.initSignedRecord(); err != nil {
+		return nil, err
+	}
+	n.SetStreamHandler(h.newStreamHandler)
+	return h, nil
+}
+
+func (h *BaseHost) initSignedRecord() error {
+	cab, ok := peerstore.GetCertifiedAddrBook(h.network.Peerstore())
+	if !ok {
+		return errors.New("peerstore does not support signed records")
+	}
+	rec := peer.PeerRecordFromAddrInfo(peer.AddrInfo{ID: h.ID(), Addrs: h.Addrs()})
```

Comment:
What happens if we _don't_ do this? I'm asking because:

1. We're going to clobber it in the Basic/Full host.
2. The addresses are going to be wrong here.

Reply:
@Stebalien Can you elaborate? It would be nice if the base host supported signed records, wouldn't it?

The addresses are only going to wrong if an address factory is used.
Also, in the full host we generate a new record for ourselves, so we'll overwrite the one that the base host generated.